examples/renderers/NMR/example2.py

```diff
@@ -0,0 +1,204 @@
+    parser.add_argument('--output_path', type=str, default=os.path.join(ROOT_DIR, 'results'),
+                        help='Path to the output directory')
+    parser.add_argument('--epochs', type=int, default=300,
+                        help='Number of epochs to optimize')
+
+    return parser.parse_args()
+
+
+class Model(nn.Module):
+
+    def __init__(self, mesh_path, image_path):
+        super(Model, self).__init__()
+
+        ###########################
+        # Load mesh
+        ###########################
+
+        mesh = TriangleMesh.from_obj(mesh_path)
+        # Normalize into unit cube, and expand such that batch size = 1
+        vertices = normalize_vertices(mesh.vertices).unsqueeze(0)
+        faces = mesh.faces.unsqueeze(0)
+
+        self.vertices = nn.Parameter(vertices)
+        self.register_buffer('faces', faces)
+
+        ###########################
+        # Generate texture (NMR format)
+        ###########################
+
+        textures = torch.ones(
+            1, self.faces.shape[1], TEXTURE_SIZE, TEXTURE_SIZE, TEXTURE_SIZE,
+            3, dtype=torch.float32
+        )
+        self.register_buffer('textures', textures)
+
+        ###########################
+        # Load target image
+        ###########################
+
+        image_ref = torch.from_numpy(imread(image_path).astype(
+            np.float32).mean(-1) / 255.)[None, ::]
+        self.register_buffer('image_ref', image_ref)
+
+        ###########################
+        # Setup renderer
+        ###########################
+
+        renderer = Renderer(camera_mode='look_at')
+        self.renderer = renderer
+
+    def forward(self):
+        ###########################
+        # Render
+        ###########################
+
+        self.renderer.eye = get_points_from_angles(
+            CAMERA_DISTANCE, ELEVATION, AZIMUTH)
+        image = self.renderer(self.vertices, self.faces, mode='silhouettes')
+        loss = torch.sum((image - self.image_ref[None, :, :]) ** 2)
+
+        return loss
+
+
+def main():
+    args = parse_arguments()
+
+    ###########################
+    # Setup model
+    ###########################
+
+    model = Model(args.mesh, args.image)
+    model.cuda()
+
+    ###########################
+    # Optimize
+    ###########################
+
+    loop = tqdm.tqdm(range(args.epochs))
+    loop.set_description('Optimizing')
+
+    optimizer = torch.optim.Adam(
+        filter(lambda p: p.requires_grad, model.parameters()))
+
+    os.makedirs(args.output_path, exist_ok=True)
+    writer = imageio.get_writer(os.path.join(
+        args.output_path, 'example2_optimization.gif'), mode='I')
+    for i in loop:
+        optimizer.zero_grad()
+
+        loss = model()
+
+        loss.backward()
+        optimizer.step()
+
+        images, _, _ = model.renderer(
+            model.vertices, model.faces, model.textures)
+
+        image = images.detach().cpu().numpy()[0].transpose(
```

Comment:
May not be a big deal but you can use `permute` on gpu before transfering to cpu